test/withDataProviders.test.js

```diff
@@ -0,0 +1,145 @@
+/* global test, expect, beforeEach, describe */
+/* eslint-disable no-console */
+import React from 'react'
+import ReactDOM from 'react-dom'
+import {withDataProviders} from '../src/withDataProviders'
+import {compose} from 'redux'
+import {connect} from 'react-redux'
+import {newTestApp, getData, getDataWithCount, GET_DATA_DELAY} from './common'
+import {Promise} from 'bluebird'
+
+beforeEach(() => {
+  getDataWithCount.counter = 0
+})
+
+test('withDataProviders (needed=true) renders child component when the data is fetched', async () => {
+  const {root} = renderMessageContainerApp({needed: true})
+
+  let renderedMessage = root.querySelector('div.message p')
+  expect(renderedMessage).toBeNull()
+
+  await Promise.delay(GET_DATA_DELAY)
+  renderedMessage = root.querySelector('div.message p')
+  expect(renderedMessage).not.toBeNull()
+  expect(renderedMessage.textContent).toBe('Hello')
+})
+
+test('withDataProviders (needed=false) renders initial state of child component', async () => {
+  const {root} = renderMessageContainerApp({needed: false, initialData: {data: 'init'}})
+
+  let renderedMessage = root.querySelector('div.message p')
+  expect(renderedMessage).not.toBeNull()
+  expect(renderedMessage.textContent).toBe('init')
+
+  await Promise.delay(GET_DATA_DELAY)
+  renderedMessage = root.querySelector('div.message p')
+  expect(renderedMessage).not.toBeNull()
+  expect(renderedMessage.textContent).toBe('Hello')
+})
+
+describe('nested withDataProviders with eager prefetching', () => {
+  const getData = [getDataWithCount, {data: ''}, GET_DATA_DELAY]
+  const MessageContainer = messageContainer({needed: true, getData, initialData: {show: false}})
+  const ParentContainer = ({show}) => (
+    <div className="container">
+      {show ? <MessageContainer /> : null}
+    </div>
+  )
+  const Container = compose(
+    connect((state) => ({show: state.show})),
+    withDataProviders(() => [messageProvider({needed: false, getData})]),
+  )(ParentContainer)
```

Comment:
For the sake of minimalistic example - you can omit the extra div in Parent Container. Just inline the function here:
```
({show}) => show && <MessageContainer />
```